history/src/history.ts

```diff
@@ -1,89 +1,96 @@
 export const closeHistorySlot = new MetaSlot<boolean>("historyClose")
 
-const historyField = new StateField({
-  init(editorState: EditorState): HistoryState {
-    return HistoryState.empty
-  },
-
-  apply(tr: Transaction, state: HistoryState, editorState: EditorState): HistoryState {
-    const fromMeta = tr.getMeta(historyStateSlot)
-    if (fromMeta) return fromMeta
-    if (tr.getMeta(closeHistorySlot)) state = state.resetTime()
-    if (!tr.changes.length && !tr.selectionSet) return state
-
-    const {newGroupDelay, minDepth} = editorState.getPluginWithField(historyField).config
-    if (tr.getMeta(MetaSlot.addToHistory) !== false)
-      return state.addChanges(tr.changes, tr.changes.length ? tr.invertedChanges() : null, tr.startState.selection,
-                              tr.getMeta(MetaSlot.time)!, tr.getMeta(MetaSlot.userEvent), newGroupDelay, minDepth)
-    return state.addMapping(tr.changes.desc, minDepth)
-  },
-
-  debugName: "historyState"
-})
-
-export function history({minDepth = 100, newGroupDelay = 500}: {minDepth?: number, newGroupDelay?: number} = {}): Plugin {
-  return new Plugin({
-    state: historyField,
-    config: {minDepth, newGroupDelay}
+function historyField(minDepth: number, newGroupDelay: number) {
+  return new StateField({
+    init(editorState: EditorState): HistoryState {
+      return HistoryState.empty
+    },
+
+    apply(tr: Transaction, state: HistoryState, editorState: EditorState): HistoryState {
+      const fromMeta = tr.getMeta(historyStateSlot)
+      if (fromMeta) return fromMeta
+      if (tr.getMeta(closeHistorySlot)) state = state.resetTime()
+      if (!tr.changes.length && !tr.selectionSet) return state
+
+      if (tr.getMeta(MetaSlot.addToHistory) !== false)
+        return state.addChanges(tr.changes, tr.changes.length ? tr.invertedChanges() : null,
+                                tr.startState.selection, tr.getMeta(MetaSlot.time)!,
+                                tr.getMeta(MetaSlot.userEvent), newGroupDelay, minDepth)
+      return state.addMapping(tr.changes.desc, minDepth)
+    },
+
+    name: "historyState"
   })
 }
 
-function historyCmd(target: PopTarget, only: ItemFilter, state: EditorState, dispatch: (tr: Transaction) => void): boolean {
-  const historyState: HistoryState | undefined = state.getField(historyField)
-  if (!historyState || !historyState.canPop(target, only)) return false
-  const {minDepth} = state.getPluginWithField(historyField).config
-  const {transaction, state: newState} = historyState.pop(target, only, state.transaction, minDepth)
-  dispatch(transaction.setMeta(historyStateSlot, newState))
-  return true
-}
+export interface HistoryConfig {minDepth?: number, newGroupDelay?: number}
 
-export function undo({state, dispatch}: {state: EditorState, dispatch: (tr: Transaction) => void}): boolean {
-  return historyCmd(PopTarget.Done, ItemFilter.OnlyChanges, state, dispatch)
-}
+class HistoryBehavior {
```

Comment:
This class is mainly a handle so that the commands and depth getters don't have to close over the configuration (and hence, field). I think the code would be easier to understand if this were either
1. split in two behaviors (`HistoryCommand`, `HistoryDepth`) or
2. reduced to only storing the two fields with the content of the methods moving down into `cmd` and `depth`.

Reply:
Pint 2. is a good idea. 1. seems like it wouldn't help much while introducing another separate element. Renaming the class to `HistoryContext` or so might cover its purpose relatively well.

Reply:
See fc5b290